File: searcher/src/main/scala/com/github/kokorin/searcher/config/ApplicationConfig.scala
package com.github.kokorin.searcher.config

import com.typesafe.config.Config
import scala.collection.JavaConverters._

trait ApplicationConfig {
  val apiConfig: ApiConfig
  val searchEngines: Seq[SearchEngineConfig]
  val aggregatorActorConfig: AggregatorActorConfig
  val searcherActorConfig: SearcherActorConfig
}

class ApplicationConfigImpl(conf: Config) extends ApplicationConfig {
  override val apiConfig: ApiConfig = new ApiConfigImpl(conf.getConfig("api"))

  override val searchEngines: Seq[SearchEngineConfig] = {
    val enginesConfig = conf.getConfig("searchEngines")
    enginesConfig.getStringList("enginesList").asScala.map { name =>
      new SearchEngineConfigImpl(enginesConfig.getConfig(name), name)
    }
  }

  private val actorsConfig = conf.getConfig("actors")

  override val aggregatorActorConfig: AggregatorActorConfig =
    new AggregatorActorConfigImpl(actorsConfig.getConfig("aggregator"))

  override val searcherActorConfig: SearcherActorConfig =
    new SearcherActorConfigImpl(actorsConfig.getConfig("searcher"))
}

File: products-reactive/src/main/kotlin/com/github/kokorin/rx/config/ApplicationConfig.kt
package com.github.kokorin.rx.config

import com.typesafe.config.Config

interface ApplicationConfig {
    val apiConfig: ApiConfig
    val databaseConfig: DatabaseConfig
}

class ApplicationConfigImpl(conf: Config): ApplicationConfig {
    override val apiConfig: ApiConfig = ApiConfigImpl(conf.getConfig("api"))
    override val databaseConfig: DatabaseConfig = DatabaseConfigImpl(conf.getConfig("database"))
}
